libtenzir/builtins/formats/lines.cpp

```diff
@@ -0,0 +1,148 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2023 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include <tenzir/argument_parser.hpp>
+#include <tenzir/detail/assert.hpp>
+#include <tenzir/plugin.hpp>
+#include <tenzir/table_slice_builder.hpp>
+#include <tenzir/to_lines.hpp>
+
+#include <optional>
+
+namespace tenzir::plugins::lines {
+
+namespace {
+
+/// The default name of the record field.
+const auto default_field_name = std::string{"data"};
+
+struct parser_args {
+  std::optional<located<std::string>> field_name;
+  std::optional<location> skip_empty;
+
+  template <class Inspector>
+  friend auto inspect(Inspector& f, parser_args& x) -> bool {
+    return f.object(x)
+      .pretty_name("parser_args")
+      .fields(f.field("field_name", x.field_name),
+              f.field("skip_empty", x.skip_empty));
+  }
+};
+
+auto line_type(std::string field_name = "data") -> type {
+  TENZIR_ASSERT(not field_name.empty());
+  return type{
+    "tenzir.line",
+    record_type{
+      {std::move(field_name), string_type{}},
+    },
+  };
+}
+
+class lines_parser final : public plugin_parser {
+public:
+  lines_parser() = default;
+
+  explicit lines_parser(parser_args args) : args_{std::move(args)} {
+  }
+
+  auto name() const -> std::string override {
+    return "lines";
+  }
+
+  auto
+  instantiate(generator<chunk_ptr> input, operator_control_plane& ctrl) const
+    -> std::optional<generator<table_slice>> override {
+    auto field_name
+      = args_.field_name ? args_.field_name->inner : default_field_name;
+    auto make
+      = [](auto& ctrl, generator<chunk_ptr> input, std::string field_name,
+           bool skip_empty) -> generator<table_slice> {
+      auto make_builder = [field_name = std::move(field_name)]() {
+        return table_slice_builder{line_type(field_name)};
+      };
+      auto num_non_empty_lines = size_t{0};
+      auto num_empty_lines = size_t{0};
+      auto builder = make_builder();
+      for (auto line : to_lines(std::move(input))) {
+        if (not line) {
+          co_yield {};
+          continue;
+        }
+        if (line->empty()) {
+          ++num_empty_lines;
+          if (skip_empty) {
+            co_yield {};
+            continue;
+          }
+        } else {
+          ++num_non_empty_lines;
+        }
+        auto num_lines = skip_empty ? num_non_empty_lines
+                                    : num_non_empty_lines + num_empty_lines;
+        if (not builder.add(*line)) {
+          diagnostic::error("failed add line")
```

Comment:
```suggestion
          diagnostic::error("failed to add line")
```

Reply:
We could also just assert in such cases. I mean, this should never fail, and we don't have to do so much bookkeeping then.

Reply:
What's the maximum line size?

Reply:
Good question. Right now, it's more or less unlimited. But that's probably fine? If the input has huge lines, this is what you want. And if it has no newlines at all, you are screwed anyway.

Reply:
So should we then keep the error handling, if it *could* fail beyond 2GB lines?

I don't think it's realistic, but the error handling is there now.

Reply:
Seems like Arrow just does a `static_cast<int32_t>(str.size())`, so... If we want to handle this, we must add some additional logic to `append_builder(...)`.

Reply:
The limit Arrow has here is actually 2 GiB per array. This applies to all arrays. Individual arrays are stored in FlatBuffers tables, which have a 32 bit offset. FlatBuffers recently added a 64-bit builder, but that'd be a breaking change in Arrow's binary format.

Reply:
Welp, I'm still missing guidance on what to actually do here now.

Reply:
Keep the error handling as-is, and fix the grammar issue. You could add an `[[unlikely]]` if you want to prematurely optimize, but that doesn't matter much.

Reply:
Agreed. Well, almost: You should remove (or truncate) the ` .hint("{}", *line)`.

Reply:
Ah, right. I removed it.